File: zy10319/standalone/WebhookServer.java
package standalone;

import com.sun.net.httpserver.HttpServer;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpExchange;

import java.io.*;
import java.net.InetSocketAddress;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

enum EventStatus {
    PENDING, WAITING, PROCESSING, SUCCESS, FAILED, TIMEOUT, SKIPPED
}

enum OutOfOrderReason {
    GAP, DUPLICATE, RETROACTIVE, TIMEOUT, NONE
}

class EventContext {
    String eventId;
    String topic;
    String businessKey;
    Long sequenceNumber;
    EventStatus status;
    OutOfOrderReason outOfOrderReason;
    String processResult;
    String errorMessage;
    LocalDateTime receivedAt;
    LocalDateTime processedAt;
    LocalDateTime waitingSince;
    Long expectedNextSequence;
}

class SequenceState {
    String topic;
    String businessKey;
    Long currentSequence;
    Long expectedNextSequence;
    Long lastProcessedSequence;
    TreeMap<Long, EventContext> waitingQueue = new TreeMap<>();
    LocalDateTime lastGapDetectedAt;
    boolean hasGap;
    Long gapStartSequence;
    
    public void init() {
        this.currentSequence = 0L;
        this.expectedNextSequence = 1L;
    }
}

class EventStore {
    private final Map<String, EventContext> eventById = new ConcurrentHashMap<>();
    private final Map<String, SequenceState> sequenceStates = new ConcurrentHashMap<>();
    
    public EventContext findEventById(String eventId) {
        return eventById.get(eventId);
    }
    
    public void saveEvent(EventContext context) {
        eventById.put(context.eventId, context);
    }
    
    public SequenceState getOrCreateSequenceState(String key, String topic, String businessKey) {
        return sequenceStates.computeIfAbsent(key, k -> {
            SequenceState state = new SequenceState();
            state.topic = topic;
            state.businessKey = businessKey;
            state.init();
            return state;
        });
    }
    
    public void saveSequenceState(String key, SequenceState state) {
        sequenceStates.put(key, state);
    }
    
    public List<SequenceState> getAllSequenceStates() {
        return new ArrayList<>(sequenceStates.values());
    }
    
    public List<EventContext> findEventsByTopicAndBusinessKey(String topic, String businessKey) {
        List<EventContext> result = new ArrayList<>();
        for (EventContext e : eventById.values()) {
            if (e.topic.equals(topic) && e.businessKey.equals(businessKey)) {
                result.add(e);
            }
        }
        Collections.sort(result, (a, b) -> a.sequenceNumber.compareTo(b.sequenceNumber));
        return result;
    }
    
    public List<EventContext> findAllEvents() {
        return new ArrayList<>(eventById.values());
    }
    
    public SequenceState findSequenceState(String topic, String businessKey) {
        String key = topic + ":" + businessKey;
        return sequenceStates.get(key);
    }
}

class SequenceService {
    private final EventStore eventStore = new EventStore();
    private final int gapTimeoutSeconds = 30;
    
    public EventStore getEventStore() {
        return eventStore;
    }
    
    public EventContext processEvent(String eventId, String topic, String businessKey, Long sequenceNumber) {
        String stateKey = topic + ":" + businessKey;
        
        EventContext existing = eventStore.findEventById(eventId);
        if (existing != null) {
            return existing;
        }
        
        SequenceState state = eventStore.getOrCreateSequenceState(stateKey, topic, businessKey);
        
        EventContext context = new EventContext();
        context.eventId = eventId;
        context.topic = topic;
        context.businessKey = businessKey;
        context.sequenceNumber = sequenceNumber;
        context.receivedAt = LocalDateTime.now();
        context.status = EventStatus.PENDING;
        context.outOfOrderReason = OutOfOrderReason.NONE;
        
        long expectedSeq = state.expectedNextSequence;
        long currentSeq = sequenceNumber;
        
        if (currentSeq == expectedSeq) {
            processInOrder(context, state);
        } else if (currentSeq < expectedSeq) {
            if (currentSeq <= state.lastProcessedSequence) {
                context.status = EventStatus.SKIPPED;
                context.outOfOrderReason = OutOfOrderReason.RETROACTIVE;
                context.errorMessage = "序列号 " + currentSeq + " 已处理过，跳过";
            } else {
                context.status = EventStatus.SKIPPED;
                context.outOfOrderReason = OutOfOrderReason.DUPLICATE;
                context.errorMessage = "序列号 " + currentSeq + " 重复，跳过";
            }
        } else {
            processOutOfOrder(context, state, expectedSeq);
        }
        
        context.expectedNextSequence = state.expectedNextSequence;
        eventStore.saveEvent(context);
        eventStore.saveSequenceState(stateKey, state);
        
        return context;
    }
    
    private void processInOrder(EventContext context, SequenceState state) {
        context.status = EventStatus.SUCCESS;
        context.processResult = "顺序处理成功";
        context.processedAt = LocalDateTime.now();
        
        state.currentSequence = context.sequenceNumber;
        state.lastProcessedSequence = context.sequenceNumber;
        state.expectedNextSequence = context.sequenceNumber + 1;
        state.hasGap = false;
        state.gapStartSequence = null;
        state.lastGapDetectedAt = null;
        
        processWaitingQueue(state);
    }
    
    private void processOutOfOrder(EventContext context, SequenceState state, long expectedSeq) {
        context.status = EventStatus.WAITING;
        context.outOfOrderReason = OutOfOrderReason.GAP;
        context.waitingSince = LocalDateTime.now();
        context.errorMessage = "序列号不连续，期望 " + expectedSeq + "，实际 " + context.sequenceNumber + "，进入等待队列";
        
        state.waitingQueue.put(context.sequenceNumber, context);
        
        if (!state.hasGap) {
            state.hasGap = true;
            state.gapStartSequence = expectedSeq;
            state.lastGapDetectedAt = LocalDateTime.now();
            System.out.println("[WARN] 检测到序列号缺口: topic=" + state.topic + 
                             ", businessKey=" + state.businessKey + 
                             ", gapStart=" + expectedSeq);
        }
    }
    
    private void processWaitingQueue(SequenceState state) {
        List<Long> processedSeqs = new ArrayList<>();
        
        for (Map.Entry<Long, EventContext> entry : state.waitingQueue.entrySet()) {
            long seq = entry.getKey();
            EventContext waitingEvent = entry.getValue();
            
            if (seq == state.expectedNextSequence) {
                waitingEvent.status = EventStatus.SUCCESS;
                waitingEvent.processResult = "从等待队列恢复处理成功";
                waitingEvent.processedAt = LocalDateTime.now();
                waitingEvent.outOfOrderReason = OutOfOrderReason.NONE;
                waitingEvent.errorMessage = null;
                
                state.currentSequence = seq;
                state.lastProcessedSequence = seq;
                state.expectedNextSequence = seq + 1;
                
                eventStore.saveEvent(waitingEvent);
                processedSeqs.add(seq);
            } else {
                break;
            }
        }
        
        for (Long seq : processedSeqs) {
            state.waitingQueue.remove(seq);
        }
        
        if (state.waitingQueue.isEmpty()) {
            state.hasGap = false;
            state.gapStartSequence = null;
            state.lastGapDetectedAt = null;
        }
    }
    
    public void checkAndProcessTimeouts() {
        List<SequenceState> allStates = eventStore.getAllSequenceStates();
        LocalDateTime now = LocalDateTime.now();
        
        for (SequenceState state : allStates) {
            if (state.hasGap && state.lastGapDetectedAt != null) {
                long secondsSinceGap = java.time.Duration.between(state.lastGapDetectedAt, now).getSeconds();
                
                if (secondsSinceGap >= gapTimeoutSeconds) {
                    processTimeout(state);
                }
            }
        }
    }
    
    private void processTimeout(SequenceState state) {
        System.out.println("[WARN] 序列号缺口超时: topic=" + state.topic + 
                         ", businessKey=" + state.businessKey + 
                         ", gapStart=" + state.gapStartSequence);

        long maxProcessedSeq = state.gapStartSequence - 1;
        
        for (Map.Entry<Long, EventContext> entry : state.waitingQueue.entrySet()) {
            EventContext waitingEvent = entry.getValue();
            
            if (waitingEvent.sequenceNumber >= state.gapStartSequence) {
                waitingEvent.status = EventStatus.SUCCESS;
                waitingEvent.processResult = "超时强制处理";
                waitingEvent.processedAt = LocalDateTime.now();
                waitingEvent.outOfOrderReason = OutOfOrderReason.TIMEOUT;
                waitingEvent.errorMessage = null;
                
                if (waitingEvent.sequenceNumber > maxProcessedSeq) {
                    maxProcessedSeq = waitingEvent.sequenceNumber;
                }
                state.lastProcessedSequence = waitingEvent.sequenceNumber;
                eventStore.saveEvent(waitingEvent);
                System.out.println("[INFO] 超时处理序列号: " + waitingEvent.sequenceNumber);
            }
        }

        state.waitingQueue.clear();
        state.hasGap = false;
        state.gapStartSequence = null;
        state.lastGapDetectedAt = null;
        state.expectedNextSequence = maxProcessedSeq + 1;
        
        eventStore.saveSequenceState(state.topic + ":" + state.businessKey, state);
    }
}

class JsonUtil {
    static String formatDateTime(LocalDateTime dt) {
        if (dt == null) return null;
        return dt.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
    }
    
    static String eventToJson(EventContext e, boolean isIdempotent) {
        StringBuilder sb = new StringBuilder();
        sb.append("{");
        sb.append("\"eventId\":\"").append(e.eventId).append("\",");
        sb.append("\"topic\":\"").append(e.topic).append("\",");
        sb.append("\"businessKey\":\"").append(e.businessKey).append("\",");
        sb.append("\"sequenceNumber\":").append(e.sequenceNumber).append(",");
        sb.append("\"status\":\"").append(e.status).append("\",");
        sb.append("\"outOfOrderReason\":\"").append(e.outOfOrderReason).append("\",");
        if (e.processResult != null) sb.append("\"processResult\":\"").append(e.processResult).append("\",");
        if (e.errorMessage != null) sb.append("\"errorMessage\":\"").append(e.errorMessage).append("\",");
        sb.append("\"receivedAt\":\"").append(formatDateTime(e.receivedAt)).append("\",");
        if (e.processedAt != null) sb.append("\"processedAt\":\"").append(formatDateTime(e.processedAt)).append("\",");
        sb.append("\"expectedNextSequence\":").append(e.expectedNextSequence).append(",");
        sb.append("\"isIdempotent\":").append(isIdempotent);
        sb.append("}");
        return sb.toString();
    }
    
    static String eventsToJson(List<EventContext> events) {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < events.size(); i++) {
            if (i > 0) sb.append(",");
            sb.append(eventToJson(events.get(i), false));
        }
        sb.append("]");
        return sb.toString();
    }
    
    static String stateToJson(SequenceState s) {
        StringBuilder sb = new StringBuilder();
        sb.append("{");
        sb.append("\"topic\":\"").append(s.topic).append("\",");
        sb.append("\"businessKey\":\"").append(s.businessKey).append("\",");
        sb.append("\"currentSequence\":").append(s.currentSequence).append(",");
        sb.append("\"expectedNextSequence\":").append(s.expectedNextSequence).append(",");
        sb.append("\"lastProcessedSequence\":").append(s.lastProcessedSequence).append(",");
        sb.append("\"waitingQueueSize\":").append(s.waitingQueue.size()).append(",");
        sb.append("\"hasGap\":").append(s.hasGap).append(",");
        if (s.gapStartSequence != null) sb.append("\"gapStartSequence\":").append(s.gapStartSequence).append(",");
        sb.append("\"waitingQueue\":").append(s.waitingQueue.keySet());
        sb.append("}");
        return sb.toString();
    }
    
    static String healthJson() {
        return "{\"status\":\"UP\",\"service\":\"webhook-sequence-api\",\"timestamp\":" + System.currentTimeMillis() + "}";
    }
    
    static String errorJson(String message) {
        return "{\"error\":\"" + message + "\"}";
    }
    
    static String parseJsonValue(String json, String key) {
        String search = "\"" + key + "\":";
        int idx = json.indexOf(search);
        if (idx < 0) return null;
        
        int start = idx + search.length();
        if (start >= json.length()) return null;
        
        while (start < json.length() && json.charAt(start) == ' ') {
            start++;
        }
        if (start >= json.length()) return null;
        
        char c = json.charAt(start);
        if (c == '"') {
            int end = json.indexOf("\"", start + 1);
            return json.substring(start + 1, end);
        } else if (Character.isDigit(c) || c == '-') {
            int end = start;
            while (end < json.length() && (Character.isDigit(json.charAt(end)) || json.charAt(end) == '.')) end++;
            return json.substring(start, end);
        }
        return null;
    }
}

public class WebhookServer {
    private static final int PORT = 8080;
    private static final SequenceService sequenceService = new SequenceService();
    
    public static void main(String[] args) throws Exception {
        HttpServer server = HttpServer.create(new InetSocketAddress(PORT), 0);
        
        server.createContext("/api/v1/events", new EventsHandler());
        server.createContext("/api/v1/events/state", new StateHandler());
        server.createContext("/api/v1/events/export", new ExportHandler());
        server.createContext("/api/v1/events/health", new HealthHandler());
        
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        
        ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
        scheduler.scheduleAtFixedRate(() -> {
            try {
                sequenceService.checkAndProcessTimeouts();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }, 5, 5, TimeUnit.SECONDS);
        
        System.out.println("========================================");
        System.out.println("  Webhook 顺序保证 API 服务已启动");
        System.out.println("  端口: " + PORT);
        System.out.println("  访问: http://localhost:" + PORT + "/api/v1/events/health");
        System.out.println("========================================");
        System.out.println();
        System.out.println("API 端点:");
        System.out.println("  POST /api/v1/events            - 提交事件");
        System.out.println("  GET  /api/v1/events/{id}       - 查询单个事件");
        System.out.println("  GET  /api/v1/events            - 查询所有事件");
        System.out.println("  GET  /api/v1/events/state      - 查询序列状态");
        System.out.println("  GET  /api/v1/events/export     - 导出所有事件");
        System.out.println("  GET  /api/v1/events/health     - 健康检查");
        System.out.println();
        System.out.println("运行测试请执行: ./test_standalone.sh");
        System.out.println();
    }
    
    static class EventsHandler implements HttpHandler {
        public void handle(HttpExchange exchange) throws IOException {
            String method = exchange.getRequestMethod();
            String path = exchange.getRequestURI().getPath();
            
            if ("POST".equals(method)) {
                handlePost(exchange);
            } else if ("GET".equals(method)) {
                handleGet(exchange, path);
            } else {
                sendResponse(exchange, 405, JsonUtil.errorJson("Method Not Allowed"));
            }
        }
        
        private void handlePost(HttpExchange exchange) throws IOException {
            try {
                InputStream is = exchange.getRequestBody();
                BufferedReader reader = new BufferedReader(new InputStreamReader(is));
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line);
                }
                String body = sb.toString();
                
                String eventId = JsonUtil.parseJsonValue(body, "eventId");
                String topic = JsonUtil.parseJsonValue(body, "topic");
                String businessKey = JsonUtil.parseJsonValue(body, "businessKey");
                String seqStr = JsonUtil.parseJsonValue(body, "sequenceNumber");
                
                if (eventId == null || topic == null || businessKey == null || seqStr == null) {
                    sendResponse(exchange, 400, JsonUtil.errorJson("Missing required fields"));
                    return;
                }
                
                Long sequenceNumber = Long.parseLong(seqStr);
                
                EventContext existing = sequenceService.getEventStore().findEventById(eventId);
                boolean isIdempotent = (existing != null);
                
                EventContext result = sequenceService.processEvent(eventId, topic, businessKey, sequenceNumber);
                
                int statusCode;
                switch (result.status) {
                    case SUCCESS: statusCode = 200; break;
                    case WAITING: statusCode = 202; break;
                    case SKIPPED: statusCode = 409; break;
                    default: statusCode = 500;
                }
                
                sendResponse(exchange, statusCode, JsonUtil.eventToJson(result, isIdempotent));
            } catch (Exception e) {
                sendResponse(exchange, 500, JsonUtil.errorJson(e.getMessage()));
            }
        }
        
        private void handleGet(HttpExchange exchange, String path) throws IOException {
            String[] parts = path.split("/");
            if (parts.length == 5 && !parts[4].isEmpty()) {
                String eventId = parts[4];
                EventContext e = sequenceService.getEventStore().findEventById(eventId);
                if (e == null) {
                    sendResponse(exchange, 404, JsonUtil.errorJson("Event not found"));
                } else {
                    sendResponse(exchange, 200, JsonUtil.eventToJson(e, false));
                }
            } else {
                List<EventContext> all = sequenceService.getEventStore().findAllEvents();
                sendResponse(exchange, 200, JsonUtil.eventsToJson(all));
            }
        }
    }
    
    static class StateHandler implements HttpHandler {
        public void handle(HttpExchange exchange) throws IOException {
            String query = exchange.getRequestURI().getQuery();
            String topic = null, businessKey = null;
            
            if (query != null) {
                for (String param : query.split("&")) {
                    String[] kv = param.split("=");
                    if (kv.length == 2) {
                        if ("topic".equals(kv[0])) topic = kv[1];
                        if ("businessKey".equals(kv[0])) businessKey = kv[1];
                    }
                }
            }
            
            if (topic == null || businessKey == null) {
                sendResponse(exchange, 400, JsonUtil.errorJson("Missing topic or businessKey"));
                return;
            }
            
            SequenceState state = sequenceService.getEventStore().findSequenceState(topic, businessKey);
            if (state == null) {
                sendResponse(exchange, 404, JsonUtil.errorJson("State not found"));
            } else {
                sendResponse(exchange, 200, JsonUtil.stateToJson(state));
            }
        }
    }
    
    static class ExportHandler implements HttpHandler {
        public void handle(HttpExchange exchange) throws IOException {
            List<EventContext> all = sequenceService.getEventStore().findAllEvents();
            String json = JsonUtil.eventsToJson(all);
            
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.getResponseHeaders().set("Content-Disposition", "attachment; filename=events.json");
            sendResponse(exchange, 200, json);
        }
    }
    
    static class HealthHandler implements HttpHandler {
        public void handle(HttpExchange exchange) throws IOException {
            sendResponse(exchange, 200, JsonUtil.healthJson());
        }
    }
    
    static void sendResponse(HttpExchange exchange, int code, String response) throws IOException {
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(code, response.getBytes("UTF-8").length);
        OutputStream os = exchange.getResponseBody();
        os.write(response.getBytes("UTF-8"));
        os.close();
    }
}

File: zy10319/webhook_server.py
#!/usr/bin/env python3
"""
Webhook 顺序保证 API - Python 版本（零依赖）
启动：python3 webhook_server.py
端口：8080
"""

import json
import sys
from datetime import datetime
from http.server import HTTPServer, BaseHTTPRequestHandler
from threading import Thread
import time

class EventStatus:
    PENDING = "PENDING"
    WAITING = "WAITING"
    PROCESSING = "PROCESSING"
    SUCCESS = "SUCCESS"
    FAILED = "FAILED"
    TIMEOUT = "TIMEOUT"
    SKIPPED = "SKIPPED"

class OutOfOrderReason:
    GAP = "GAP"
    DUPLICATE = "DUPLICATE"
    RETROACTIVE = "RETROACTIVE"
    TIMEOUT = "TIMEOUT"
    NONE = "NONE"

class EventContext:
    def __init__(self, event_id, topic, business_key, sequence_number):
        self.event_id = event_id
        self.topic = topic
        self.business_key = business_key
        self.sequence_number = sequence_number
        self.status = EventStatus.PENDING
        self.out_of_order_reason = OutOfOrderReason.NONE
        self.process_result = None
        self.error_message = None
        self.received_at = datetime.now()
        self.processed_at = None
        self.expected_next_sequence = None

    def to_dict(self, is_idempotent=False):
        return {
            "eventId": self.event_id,
            "topic": self.topic,
            "businessKey": self.business_key,
            "sequenceNumber": self.sequence_number,
            "status": self.status,
            "outOfOrderReason": self.out_of_order_reason,
            "processResult": self.process_result,
            "errorMessage": self.error_message,
            "receivedAt": self.received_at.isoformat() if self.received_at else None,
            "processedAt": self.processed_at.isoformat() if self.processed_at else None,
            "expectedNextSequence": self.expected_next_sequence,
            "isIdempotent": is_idempotent
        }

class SequenceState:
    def __init__(self, topic, business_key):
        self.topic = topic
        self.business_key = business_key
        self.current_sequence = 0
        self.expected_next_sequence = 1
        self.last_processed_sequence = None
        self.waiting_queue = {}
        self.last_gap_detected_at = None
        self.has_gap = False
        self.gap_start_sequence = None

    def to_dict(self):
        return {
            "topic": self.topic,
            "businessKey": self.business_key,
            "currentSequence": self.current_sequence,
            "expectedNextSequence": self.expected_next_sequence,
            "lastProcessedSequence": self.last_processed_sequence,
            "hasGap": self.has_gap,
            "gapStartSequence": self.gap_start_sequence,
            "waitingQueue": list(self.waiting_queue.keys()),
            "waitingQueueSize": len(self.waiting_queue)
        }

class SequenceService:
    def __init__(self):
        self.event_by_id = {}
        self.sequence_states = {}
        self.gap_timeout_seconds = 30

    def get_state_key(self, topic, business_key):
        return f"{topic}:{business_key}"

    def get_or_create_state(self, topic, business_key):
        key = self.get_state_key(topic, business_key)
        if key not in self.sequence_states:
            self.sequence_states[key] = SequenceState(topic, business_key)
        return self.sequence_states[key]

    def process_event(self, event_id, topic, business_key, sequence_number):
        existing = self.event_by_id.get(event_id)
        if existing:
            return existing, True
        
        state = self.get_or_create_state(topic, business_key)
        context = EventContext(event_id, topic, business_key, sequence_number)
        
        expected_seq = state.expected_next_sequence
        current_seq = sequence_number
        
        if current_seq == expected_seq:
            self._process_in_order(context, state)
        elif current_seq < expected_seq:
            if state.last_processed_sequence and current_seq <= state.last_processed_sequence:
                context.status = EventStatus.SKIPPED
                context.out_of_order_reason = OutOfOrderReason.RETROACTIVE
                context.error_message = f"序列号 {current_seq} 已处理过，跳过"
            else:
                context.status = EventStatus.SKIPPED
                context.out_of_order_reason = OutOfOrderReason.DUPLICATE
                context.error_message = f"序列号 {current_seq} 重复，跳过"
        else:
            self._process_out_of_order(context, state, expected_seq)
        
        context.expected_next_sequence = state.expected_next_sequence
        self.event_by_id[event_id] = context
        
        return context, False

    def _process_in_order(self, context, state):
        context.status = EventStatus.SUCCESS
        context.process_result = "顺序处理成功"
        context.processed_at = datetime.now()
        
        state.current_sequence = context.sequence_number
        state.last_processed_sequence = context.sequence_number
        state.expected_next_sequence = context.sequence_number + 1
        state.has_gap = False
        state.gap_start_sequence = None
        state.last_gap_detected_at = None
        
        self._process_waiting_queue(state)

    def _process_out_of_order(self, context, state, expected_seq):
        context.status = EventStatus.WAITING
        context.out_of_order_reason = OutOfOrderReason.GAP
        context.error_message = f"序列号不连续，期望 {expected_seq}，实际 {context.sequence_number}，进入等待队列"
        
        state.waiting_queue[context.sequence_number] = context
        
        if not state.has_gap:
            state.has_gap = True
            state.gap_start_sequence = expected_seq
            state.last_gap_detected_at = datetime.now()
            print(f"[WARN] 检测到序列号缺口: topic={state.topic}, businessKey={state.business_key}, gapStart={expected_seq}")

    def _process_waiting_queue(self, state):
        processed_seqs = []
        
        for seq in sorted(state.waiting_queue.keys()):
            waiting_event = state.waiting_queue[seq]
            if seq == state.expected_next_sequence:
                waiting_event.status = EventStatus.SUCCESS
                waiting_event.process_result = "从等待队列恢复处理成功"
                waiting_event.processed_at = datetime.now()
                waiting_event.out_of_order_reason = OutOfOrderReason.NONE
                waiting_event.error_message = None
                
                state.current_sequence = seq
                state.last_processed_sequence = seq
                state.expected_next_sequence = seq + 1
                
                processed_seqs.append(seq)
            else:
                break
        
        for seq in processed_seqs:
            del state.waiting_queue[seq]
        
        if not state.waiting_queue:
            state.has_gap = False
            state.gap_start_sequence = None
            state.last_gap_detected_at = None

    def process_timeout(self, topic, business_key):
        state = self.get_or_create_state(topic, business_key)
        if not state.has_gap:
            return []
        
        print(f"[WARN] 序列号缺口超时: topic={topic}, businessKey={business_key}, gapStart={state.gap_start_sequence}")
        
        processed = []
        max_processed_seq = state.gap_start_sequence - 1
        
        for seq in sorted(state.waiting_queue.keys()):
            waiting_event = state.waiting_queue[seq]
            
            if waiting_event.sequence_number >= state.gap_start_sequence:
                waiting_event.status = EventStatus.SUCCESS
                waiting_event.process_result = "超时强制处理"
                waiting_event.processed_at = datetime.now()
                waiting_event.out_of_order_reason = OutOfOrderReason.TIMEOUT
                waiting_event.error_message = None
                
                if waiting_event.sequence_number > max_processed_seq:
                    max_processed_seq = waiting_event.sequence_number
                
                state.last_processed_sequence = waiting_event.sequence_number
                processed.append(waiting_event)
                print(f"[INFO] 超时处理序列号: {seq}")
        
        state.waiting_queue.clear()
        state.has_gap = False
        state.gap_start_sequence = None
        state.last_gap_detected_at = None
        state.expected_next_sequence = max_processed_seq + 1
        
        return processed

    def check_timeouts(self):
        now = datetime.now()
        for state in self.sequence_states.values():
            if state.has_gap and state.last_gap_detected_at:
                elapsed = (now - state.last_gap_detected_at).total_seconds()
                if elapsed >= self.gap_timeout_seconds:
                    self.process_timeout(state.topic, state.business_key)

service = SequenceService()

def timeout_checker():
    while True:
        service.check_timeouts()
        time.sleep(5)

class WebhookHandler(BaseHTTPRequestHandler):
    def _set_headers(self, status_code=200, content_type="application/json"):
        self.send_response(status_code)
        self.send_header("Content-Type", content_type)
        self.end_headers()

    def _send_json(self, data, status_code=200):
        self._set_headers(status_code)
        self.wfile.write(json.dumps(data, ensure_ascii=False, indent=2).encode("utf-8"))

    def do_GET(self):
        path = self.path
        
        if path == "/api/v1/events/health":
            self._send_json({
                "status": "UP",
                "service": "webhook-sequence-api",
                "timestamp": int(time.time() * 1000)
            })
            return
        
        if path.startswith("/api/v1/events/state"):
            query = path.split("?")[-1] if "?" in path else ""
            params = {}
            for param in query.split("&"):
                if "=" in param:
                    k, v = param.split("=", 1)
                    params[k] = v
            
            if "topic" in params and "businessKey" in params:
                state = service.get_or_create_state(params["topic"], params["businessKey"])
                self._send_json(state.to_dict())
                return
            else:
                self._send_json({"error": "Missing topic or businessKey"}, 400)
                return
        
        if path.startswith("/api/v1/events/export"):
            query = path.split("?")[-1] if "?" in path else ""
            params = {}
            for param in query.split("&"):
                if "=" in param:
                    k, v = param.split("=", 1)
                    params[k] = v
            
            events = list(service.event_by_id.values())
            
            if "topic" in params and "businessKey" in params:
                events = [e for e in events if e.topic == params["topic"] and e.business_key == params["businessKey"]]
            elif "topic" in params:
                events = [e for e in events if e.topic == params["topic"]]
            elif "businessKey" in params:
                events = [e for e in events if e.business_key == params["businessKey"]]
            
            self._send_json([e.to_dict() for e in events])
            return
        
        if path.startswith("/api/v1/events/trigger-timeout"):
            query = path.split("?")[-1] if "?" in path else ""
            params = {}
            for param in query.split("&"):
                if "=" in param:
                    k, v = param.split("=", 1)
                    params[k] = v
            
            if "topic" in params and "businessKey" in params:
                processed = service.process_timeout(params["topic"], params["businessKey"])
                self._send_json({
                    "processed": len(processed),
                    "events": [e.to_dict() for e in processed]
                })
                return
            else:
                self._send_json({"error": "Missing topic or businessKey"}, 400)
                return
        
        if path.startswith("/api/v1/events/"):
            event_id = path.split("/")[-1]
            if event_id in service.event_by_id:
                self._send_json(service.event_by_id[event_id].to_dict())
                return
            else:
                self._send_json({"error": "Event not found"}, 404)
                return
        
        if path == "/api/v1/events" or path == "/api/v1/events/":
            events = [e.to_dict() for e in service.event_by_id.values()]
            self._send_json(events)
            return
        
        self._send_json({"error": "Not found"}, 404)

    def do_POST(self):
        path = self.path
        
        if path == "/api/v1/events" or path == "/api/v1/events/":
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length).decode("utf-8")
            
            try:
                data = json.loads(body)
            except json.JSONDecodeError:
                self._send_json({"error": "Invalid JSON"}, 400)
                return
            
            required = ["eventId", "topic", "businessKey", "sequenceNumber"]
            for field in required:
                if field not in data:
                    self._send_json({"error": f"Missing required field: {field}"}, 400)
                    return
            
            event, is_idempotent = service.process_event(
                data["eventId"],
                data["topic"],
                data["businessKey"],
                int(data["sequenceNumber"])
            )
            
            if event.status == EventStatus.SUCCESS:
                status_code = 200
            elif event.status == EventStatus.WAITING:
                status_code = 202
            elif event.status == EventStatus.SKIPPED:
                status_code = 409
            else:
                status_code = 500
            
            self._send_json(event.to_dict(is_idempotent), status_code)
            return
        
        self._send_json({"error": "Not found"}, 404)

    def log_message(self, format, *args):
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] {args[0]}")

def run(port=8080):
    server = HTTPServer(("0.0.0.0", port), WebhookHandler)
    
    Thread(target=timeout_checker, daemon=True).start()
    
    print("=" * 60)
    print("  Webhook 顺序保证 API - Python 服务器已启动")
    print(f"  端口: {port}")
    print("=" * 60)
    print()
    print("API 端点:")
    print("  POST /api/v1/events            - 提交事件")
    print("  GET  /api/v1/events/{id}       - 查询单个事件")
    print("  GET  /api/v1/events            - 查询所有事件")
    print("  GET  /api/v1/events/state      - 查询序列状态")
    print("  GET  /api/v1/events/export     - 导出所有事件")
    print("  GET  /api/v1/events/health     - 健康检查")
    print()
    print("测试命令:")
    print("  curl http://localhost:8080/api/v1/events/health")
    print("  curl -X POST http://localhost:8080/api/v1/events \\")
    print("    -H 'Content-Type: application/json' \\")
    print("    -d '{\"eventId\":\"evt-1\",\"topic\":\"order\",\"businessKey\":\"biz-1\",\"sequenceNumber\":1}'")
    print()
    
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n服务器已停止")
        server.server_close()

if __name__ == "__main__":
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8080
    run(port)
